master/internal/stream/projects.go

```diff
@@ -0,0 +1,226 @@
+)
+
+const (
+	// ProjectsDeleteKey specifies the key for delete projects.
+	ProjectsDeleteKey = "projects_deleted"
+	// ProjectsUpsertKey specifies the key for upsert projects.
+	ProjectsUpsertKey = "project"
+)
+
+// ProjectMsg is a stream.Msg.
+type ProjectMsg struct {
+	bun.BaseModel `bun:"table:projects"`
+
+	// immutable attributes
+	ID int `bun:"id,pk" json:"id"`
+
+	// mutable attributes
+	Name        string      `bun:"name" json:"name"`
+	Description string      `bun:"description" json:"description"`
+	Archived    bool        `bun:"archived" json:"archived"`
+	CreatedAt   time.Time   `bun:"created_at" json:"created_at"`
+	Notes       JSONB       `bun:"notes" json:"notes"`
+	WorkspaceID int         `bun:"workspace_id" json:"workspace_id"`
+	UserID      int         `bun:"user_id" json:"user_id"`
+	Immutable   bool        `bun:"immutable" json:"immutable"`
+	State       model.State `bun:"state" json:"state"`
+
+	// metadata
+	Seq int64 `bun:"seq" json:"seq"`
+}
+
+// SeqNum gets the SeqNum from a ProjectMsg.
+func (pm *ProjectMsg) SeqNum() int64 {
+	return pm.Seq
+}
+
+// UpsertMsg creates a Project stream upsert message.
+func (pm *ProjectMsg) UpsertMsg() stream.UpsertMsg {
+	return stream.UpsertMsg{
+		JSONKey: ProjectsUpsertKey,
+		Msg:     pm,
+	}
+}
+
+// DeleteMsg creates a Project stream delete message.
+func (pm *ProjectMsg) DeleteMsg() stream.DeleteMsg {
+	deleted := strconv.FormatInt(int64(pm.ID), 10)
+	return stream.DeleteMsg{
+		Key:     ProjectsDeleteKey,
+		Deleted: deleted,
+	}
+}
+
+// ProjectSubscriptionSpec is what a user submits to define a project subscription.
+type ProjectSubscriptionSpec struct {
+	WorkspaceIDs []int `json:"workspace_ids"`
+	ProjectIDs   []int `json:"project_ids"`
+	Since        int64 `json:"since"`
+}
+
+// ProjectCollectStartupMsgs collects ProjectMsg's that were missed prior to startup.
+func ProjectCollectStartupMsgs(
+	ctx context.Context,
+	user model.User,
+	known string,
+	spec ProjectSubscriptionSpec,
+) (
+	[]stream.PreparableMessage, error,
+) {
+	var out []stream.PreparableMessage
+
+	if len(spec.WorkspaceIDs) == 0 && len(spec.ProjectIDs) == 0 {
+		// empty subscription: everything known should be returned as deleted
+		out = append(out, stream.DeleteMsg{
+			Key:     ProjectsDeleteKey,
+			Deleted: known,
+		})
+		return out, nil
+	}
+	// step 0: get user's permitted access scopes
+	accessMap, err := AuthZProvider.Get().GetProjectStreamableScopes(ctx, user)
+	if err != nil {
+		return nil, err
+	}
+	var accessScopes []model.AccessScopeID
+	for id, isPermitted := range accessMap {
+		if isPermitted {
+			accessScopes = append(accessScopes, id)
+		}
+	}
+
+	// step 1: calculate all ids matching this subscription
+	var data []*ProjectMsg
+	q := db.Bun().NewSelect().Model(&data).Order("id ASC")
+	q = permFilter(q, accessMap, accessScopes)
+
+	// Ignore tmf.Since, because we want appearances, which might not be have seq > spec.Since.
+	ws := stream.WhereSince{Since: 0}
```

Comment:
The appearance case that we're specifically trying to avoid here is
- when a permission scope change results in the entity falling into scope without causing a update to the row; however, this issue is avoided if we have a trigger that will chain down permission scope changes...that sounds like a stampede

Reply:
I'm questioning if this is an issue. If a since value is specified, the user is not interested in receiving any information prior to that seq value (or timestamp), regardless of whether an entity now appears in their permission scope. 

The expected behavior should be if a change occurs to the now-in-scope entity that results in the publisher receiving an event, and then streamer should receive that message. The intention in this function should be to capture relevant events that occurred while the streamer was not connected, and the since value should be considered as part of how that scope is defined. 

This further supports using a timestamp due to the complications that could arise from overflow.

My concerns with this behavior would be if that entity got deleted as the first event this streamer was aware of it's existence it would potentially appear clunky. That being said, I would argue that this isn't where we should be trying to solve this problem, that after the client creates a bulk of a startup message and prior to sending it to the streaming server, the relevant endpoint should be hit to create set of known entities filtered based the subscription excluding than since value,  instead of ignoring the since value all together.

Reply:
I think it's correct as written.  I want to make a couple comments about questions you asked:

> If a since value is specified, the user is not interested in receiving any information prior to that seq value (or timestamp), regardless of whether an entity now appears in their permission scope.

Not quite.  `since` is about efficient streaming of differences from when the client last checked in until now.  Appearances are absolutely relevant.  Imagine a webui that is rendering a list of experiments and saving them to local storage.  A user shows up and opens the webui, and the webui wants to, as efficiently as possible, update what it has stored in the cache with fresh information before rendering it.  If an appearance event occurred since the client last connected, there will be new experiments that the client can see, even if those experiments have a since value which is less than what the client connected with.

So in order for the server to detect those appearance cases, the server has to query the database with no since value to get the full set of primary keys, which can then be compared against the client's `known` values to detect appearances.

Reply:
Okay, I think I see my confusion. It doesn't look like we are using the `since` value, outside of this, it's less that this is incorrect, and more than we should actually be including a check against the since value in the filter while live streaming

Reply:
hm... you are right.  This must be a bug.

Though the bug can't be about filtering the online events... the client really has no reason to ever state a since value that it hasn't actually seen, so I am fine if we don't filter honor since for those live-streaming events.

The bug is that a client-known object with updated so that seq > since, the upsert message would not be sent as part of the initial sync messages.

I think this was true in my original prototype as well.  I think since I was selecting everything ignoring seq, I was grabbing those IDs from the database, so I thought I had it covered, but I was not actually sending the upserts to the client.